File: rust/openmso/src/transport.rs
// SPDX-License-Identifier: Apache-2.0
//! nng sockets carrying protobuf: REQ/REP for control, PUSH/PULL for events.

use std::path::{Path, PathBuf};
use std::time::Duration;

use nng::options::{protocol::reqrep::ResendTime, Options, RecvMaxSize, RecvTimeout, SendTimeout};
use nng::{Protocol, Socket};

use crate::{Error, Result};

/// Largest `CaptureData.payload` a plugin should emit. The protocol's ceiling
/// is 4 MiB; staying well under it leaves room for run-length encoding, whose
/// worst case expands a chunk by one byte per sample.
pub const MAX_PAYLOAD: usize = 1 << 20;

/// Applied to every socket rather than trusted to the default, per the spike.
pub const RECV_MAX_SIZE: usize = 8 << 20;

/// Build a socket with the options every OCP endpoint needs.
pub fn socket(protocol: Protocol) -> Result<Socket> {
    let s = Socket::new(protocol)?;
    s.set_opt::<RecvMaxSize>(RECV_MAX_SIZE)?;
    s.set_opt::<SendTimeout>(Some(Duration::from_secs(30)))?;
    if protocol == Protocol::Req0 {
        // nng re-sends a request whose reply is slower than this timer, and
        // the far end runs it a second time — measured, not theoretical. On a
        // local socket the pipe either exists or breaks, so resend buys
        // nothing; zero disables it.
        s.set_opt::<ResendTime>(Some(Duration::ZERO))?;
    }
    Ok(s)
}

/// Listen on `url`, restricting an `ipc://` socket to the current user.
///
/// `Socket::listen` starts the listener immediately, and IPC permissions can
/// only be set before it starts, so this goes through a builder.
pub fn listen(socket: &Socket, url: &str) -> Result<nng::Listener> {
    let builder = nng::ListenerBuilder::new(socket, url)?;
    #[cfg(unix)]
    if url.starts_with("ipc://") {
        use nng::options::transport::ipc::Permissions;
        builder.set_opt::<Permissions>(0o600)?;
    }
    builder.start().map_err(|(_, e)| Error::Nng(e))
}

pub fn set_recv_timeout(socket: &Socket, timeout: Option<Duration>) -> Result<()> {
    socket.set_opt::<RecvTimeout>(timeout).map_err(Error::from)
}

pub fn send<M: prost::Message>(socket: &Socket, message: &M) -> Result<()> {
    let mut buf = Vec::with_capacity(message.encoded_len());
    message.encode(&mut buf).expect("Vec never runs out of room");
    socket.send(&buf[..]).map_err(|(_, e)| Error::Nng(e))
}

pub fn recv<M: prost::Message + Default>(socket: &Socket) -> Result<M> {
    let msg = socket.recv()?;
    M::decode(msg.as_slice()).map_err(Error::from)
}

/// A directory holding one plugin process's two sockets, removed on drop.
///
/// The frontend owns this; the plugin only ever sees the two URLs. Windows has
/// no directory to make — `ipc://` there is a named pipe in a flat namespace.
pub struct Endpoints {
    pub control: String,
    pub events: String,
    #[cfg(unix)]
    dir: PathBuf,
}

impl Endpoints {
    pub fn new() -> Result<Self> {
        // Four hex digits of process-and-clock entropy. sun_path is 108 bytes
        // on Linux and 104 on macOS, and a long path is rejected outright, so
        // the random component has to stay short.
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.subsec_nanos())
            .unwrap_or(0);
        let mut tag = (now ^ std::process::id()) & 0xffff;
        for _ in 0..64 {
            match Self::create(tag) {
                Err(Error::Io(e)) if e.kind() == std::io::ErrorKind::AlreadyExists => {
                    tag = (tag + 1) & 0xffff;
                }
                other => return other,
            }
        }
        Err(Error::Protocol("no free runtime directory".into()))
    }

    #[cfg(unix)]
    fn create(tag: u32) -> Result<Self> {
        use std::os::unix::fs::{DirBuilderExt, PermissionsExt};

        let base = std::env::var_os("XDG_RUNTIME_DIR")
            .map(PathBuf::from)
            .unwrap_or_else(std::env::temp_dir)
            .join("openmso");
        std::fs::DirBuilder::new().recursive(true).mode(0o700).create(&base)?;
        // A pre-existing directory keeps its old mode, and it holds every
        // running plugin's sockets.
        std::fs::set_permissions(&base, std::fs::Permissions::from_mode(0o700))?;

        let dir = base.join(format!("{tag:04x}"));
        std::fs::DirBuilder::new().mode(0o700).create(&dir)?;

        let control = dir.join("ctl");
        check_sun_path(&control)?;
        Ok(Endpoints {
            control: format!("ipc://{}", control.display()),
            events: format!("ipc://{}", dir.join("evt").display()),
            dir,
        })
    }

    #[cfg(windows)]
    fn create(tag: u32) -> Result<Self> {
        Ok(Endpoints {
            control: format!("ipc://openmso-{tag:04x}-ctl"),
            events: format!("ipc://openmso-{tag:04x}-evt"),
        })
    }
}

#[cfg(unix)]
impl Drop for Endpoints {
    fn drop(&mut self) {
        std::fs::remove_dir_all(&self.dir).ok();
    }
}

/// macOS allows 104 bytes including the terminator; budget against the tighter
/// of the two so a capture that works here works there.
#[cfg(unix)]
fn check_sun_path(path: &Path) -> Result<()> {
    const LIMIT: usize = 103;
    let len = path.as_os_str().len();
    if len > LIMIT {
        return Err(Error::Protocol(format!(
            "socket path is {len} bytes, over the {LIMIT}-byte limit: {}",
            path.display()
        )));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn endpoints_are_private_and_vanish_with_the_frontend() {
        use std::os::unix::fs::PermissionsExt;

        let dir = {
            let e = Endpoints::new().unwrap();
            assert!(e.control.starts_with("ipc://"));
            assert_ne!(e.control, e.events);
            let mode = std::fs::metadata(&e.dir).unwrap().permissions().mode();
            assert_eq!(mode & 0o777, 0o700, "runtime dir must not be world-readable");
            e.dir.clone()
        };
        assert!(!dir.exists(), "runtime directory outlived its Endpoints");
    }

    #[test]
    fn concurrent_frontends_get_separate_directories() {
        let (a, b) = (Endpoints::new().unwrap(), Endpoints::new().unwrap());
        assert_ne!(a.control, b.control);
    }

    #[test]
    fn overlong_socket_paths_are_rejected_before_nng_sees_them() {
        let long = PathBuf::from("/run/user/1000").join("x".repeat(100));
        assert!(check_sun_path(&long).is_err());
        assert!(check_sun_path(Path::new("/run/user/1000/openmso/7f3a/ctl")).is_ok());
    }
}

File: cpp/src/transport.cpp
// SPDX-License-Identifier: Apache-2.0

#include "openmso/transport.h"

#include <nng/protocol/pipeline0/pull.h>
#include <nng/protocol/pipeline0/push.h>
#include <nng/protocol/reqrep0/rep.h>
#include <nng/protocol/reqrep0/req.h>

#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include <cerrno>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>

namespace openmso::transport {

namespace {

constexpr nng_duration SEND_TIMEOUT_MS = 30'000;

int openSocket(nng_socket *s, Protocol protocol)
{
    switch (protocol) {
    case Protocol::Req0:  return nng_req0_open(s);
    case Protocol::Rep0:  return nng_rep0_open(s);
    case Protocol::Push0: return nng_push0_open(s);
    case Protocol::Pull0: return nng_pull0_open(s);
    }
    return NNG_EINVAL;
}

void check(int rv, const std::string &context)
{
    if (rv != 0)
        throw Error::nng(rv, context);
}

} // namespace

Socket::Socket(Protocol protocol)
{
    check(openSocket(&socket_, protocol), "open");
    open_ = true;

    try {
        check(nng_socket_set_size(socket_, NNG_OPT_RECVMAXSZ, RECV_MAX_SIZE),
              "set recv-size-max");
        check(nng_socket_set_ms(socket_, NNG_OPT_SENDTIMEO, SEND_TIMEOUT_MS),
              "set send-timeout");
        if (protocol == Protocol::Req0) {
            // nng re-sends a request whose reply outlasts this timer, and the
            // far end runs it twice. Resend buys nothing on a local socket.
            check(nng_socket_set_ms(socket_, NNG_OPT_REQ_RESENDTIME, 0),
                  "set req:resend-time");
        }
    } catch (...) {
        close();
        throw;
    }
}

Socket::~Socket() { close(); }

Socket::Socket(Socket &&other) noexcept
    : socket_(other.socket_), open_(other.open_)
{
    other.open_ = false;
}

Socket &Socket::operator=(Socket &&other) noexcept
{
    if (this != &other) {
        close();
        socket_ = other.socket_;
        open_ = other.open_;
        other.open_ = false;
    }
    return *this;
}

void Socket::close()
{
    if (open_) {
        nng_close(socket_);
        open_ = false;
    }
}

void Socket::listen(const std::string &url)
{
    // Not nng_listen: IPC permissions must be set before the listener starts.
    nng_listener listener{};
    check(nng_listener_create(&listener, socket_, url.c_str()), "listener create");

#ifndef _WIN32
    if (url.rfind("ipc://", 0) == 0) {
        const int rv =
            nng_listener_set_int(listener, NNG_OPT_IPC_PERMISSIONS, 0600);
        if (rv != 0) {
            nng_listener_close(listener);
            throw Error::nng(rv, "set ipc:permissions");
        }
    }
#endif

    const int rv = nng_listener_start(listener, 0);
    if (rv != 0) {
        nng_listener_close(listener);
        throw Error::nng(rv, "listen on " + url);
    }
}

void Socket::dial(const std::string &url)
{
    check(nng_dial(socket_, url.c_str(), nullptr, 0), "dial " + url);
}

void Socket::setRecvTimeout(std::optional<std::chrono::milliseconds> timeout)
{
    const nng_duration d =
        timeout ? static_cast<nng_duration>(timeout->count()) : NNG_DURATION_INFINITE;
    check(nng_socket_set_ms(socket_, NNG_OPT_RECVTIMEO, d), "set recv-timeout");
}

void Socket::send(const google::protobuf::MessageLite &message)
{
    std::string buf;
    if (!message.SerializeToString(&buf))
        throw Error::protocol("could not serialize " + message.GetTypeName());

    check(nng_send(socket_, buf.data(), buf.size(), 0), "send");
}

void Socket::recv(google::protobuf::MessageLite &out)
{
    void *data = nullptr;
    std::size_t size = 0;
    check(nng_recv(socket_, &data, &size, NNG_FLAG_ALLOC), "recv");

    const bool ok = out.ParseFromArray(data, static_cast<int>(size));
    nng_free(data, size);
    if (!ok)
        throw Error::decode("could not parse " + out.GetTypeName());
}

Endpoints::Endpoints()
{
    // Four hex digits only: sun_path is short and hard-limited.
    struct timespec ts {};
    clock_gettime(CLOCK_REALTIME, &ts);
    unsigned tag = (static_cast<unsigned>(ts.tv_nsec) ^
                    static_cast<unsigned>(::getpid())) & 0xffffu;

    const char *runtime = std::getenv("XDG_RUNTIME_DIR");
    const char *tmp = std::getenv("TMPDIR");
    std::string base = runtime ? runtime : (tmp ? tmp : "/tmp");
    base += "/openmso";

    if (::mkdir(base.c_str(), 0700) != 0 && errno != EEXIST)
        throw Error::io(errno, "create " + base);
    // A pre-existing directory keeps its old mode, and it holds every running
    // plugin's sockets.
    if (::chmod(base.c_str(), 0700) != 0)
        throw Error::io(errno, "chmod " + base);

    for (int attempt = 0; attempt < 64; ++attempt) {
        char name[8];
        std::snprintf(name, sizeof(name), "%04x", tag);
        const std::string dir = base + "/" + name;

        if (::mkdir(dir.c_str(), 0700) != 0) {
            if (errno == EEXIST) {
                tag = (tag + 1) & 0xffffu;
                continue;
            }
            throw Error::io(errno, "create " + dir);
        }

        const std::string control = dir + "/ctl";
        try {
            checkSunPath(control);
        } catch (...) {
            ::rmdir(dir.c_str());
            throw;
        }

        dir_ = dir;
        control_ = "ipc://" + control;
        events_ = "ipc://" + dir + "/evt";
        return;
    }

    throw Error::protocol("no free runtime directory");
}

Endpoints::~Endpoints()
{
    if (dir_.empty())
        return;
    ::unlink((dir_ + "/ctl").c_str());
    ::unlink((dir_ + "/evt").c_str());
    ::rmdir(dir_.c_str());
}

void checkSunPath(const std::string &path)
{
    constexpr std::size_t LIMIT = 103;
    if (path.size() > LIMIT) {
        throw Error::protocol("socket path is " + std::to_string(path.size()) +
                              " bytes, over the " + std::to_string(LIMIT) +
                              "-byte limit: " + path);
    }
}

} // namespace openmso::transport
